### DISCUSSION TITLE
NBT compounds versus native Skript performance

### DISCUSSION THREAD

**Discussion by SandsByte** [ShaneBeee/SkBee#281] [Q&A] [answered]
Firstly, I love bees and I also love SkBee. I really appreciate this project and it has been so helpful to unlock important constraints in Skript. Some of which (arguably) should be included with Skript by default, but I digress.

In the process of writing a very ambitious script. My script is presently suffering from long load times due to large indexes. Which is another issue entirely.
Overall, I'm concerned about performance but don't exactly know how to benchmark it.

Example 1:
Players basically break blocks other players haven't placed and get points for them. To check if they've been placed, I added a simple NBT custom tag to certain blocks placed. And... of course a simple check for the tag when they've been broken.

Q: Is it more performant to use coordinates in a variable or checking for an NBT tag when a block is broken?

Essentially this
```
on break:
    if {placed::%player%::*} doesn't contain event-location:
        add 10 to {_points}
    else:
        add event-location to {placed::%player%::*}
```
Versus
```
on place:
    set {_n} to nbt compound of event-block
    set tag "custom;placed" of {_n} to 1
    set nbt-block at event-location to event-block with nbt {_c}
on break:
    nbt compound of event-block doesn't have tag "custom;placed"
    add 10 points to {_points}
```

My first thought is the NBT solution is much quicker because it doesn't need to cross-check a potentially giant variable. On the other hand, how the storage and check of NBT tags for each placed block could also cause some performance issue in addition to nbt-block function...?

Thank you so much in advance for any insight

**Comment by ShaneBeee** [comment, marked as answer]
for starters:
```vb
on place:
    set {_n} to nbt compound of event-block
    set tag "custom;placed" of {_n} to 1
    set nbt-block at event-location to event-block with nbt {_c}
```
This is overkill, you just need to set the tag, ex:
```vb
on place:
    set tag "custom; placed" of nbt compound of event-block to 1
```
Simple

As for the rest of your question, NBT is quite fast to the best of my knowledge.
It's a bit hard to do bench marking on this since the actual process is quite fast. 

When it comes to storage, having thousands of variables just to know if a player placed a block is serious overkill and WILL harm your server, since when the server loads, it loads EVERY VARIABLE into ram,
in this scenario, I'd highly recommend using NBT

**Reply by SandsByte** [reply]
Ah, so similarly to "on pick up", the "on place" is triggered before the block is actually placed. There's no need to put a new block. That is much more elegant.
And so far, the NBT seems to be working really well. There's already significant load times with a few hundred variables... so the less of those the better. Thank you
